doc/sfptpdctl.8
.Dd April 09, 2023
.Os Linux
.Dt SFPTPDCTL 8 SMM
.Sh NAME
.Nm sfptpdctl ,
.Nm sfptpdctl.py
.Nd sfptpd control client
.Sh SYNOPSIS
.Nm sfptpdctl
.Op Ar command Ns Op = Ns Ar parameter
.Nm sfptpdctl.py
.Ar command Ns Op = Ns Ar parameter
.Sh DESCRIPTION
Issue a control command to the sfptpd daemon.
.Pp
This utility uses the control interface for sfptpd over a Unix Domain Socket.
.Pp
The control interface is a unidirectional public interface that system administration scripts may use directly and an example Python script is provided to demonstrate this usage.
.Sh FILES
.Bl -tag -width /usr/share/doc/sfptpd/examples/sfptpdctl.py
.It Pa /run/sfptpd/control-v1.sock
the sfptpd control socket
.It Pa /usr/share/doc/sfptpd/examples/sfptpdctl.c
source code for
.Nm sfptpdctl
.It Pa /ust/share/doc/sfptpd/examples/sfptpdctl.py
example script for using the sfptpd control interface
.El
.Sh EXAMPLES
.Ss Log rotation
The following command will cause log files to be closed and opened to enable system log rotation.
.Pp
.Dl # sfptpdctl logrotate
.Ss Manual sync instance selection
When manual sync instance selection has been enabled with a configuration such as:
.Pp
.D1 selection_policy manual ptp1
.Pp
the following command will cause an alternative sync instance to be selected:
.Pp
.Dl # sfptpdctl selectinstance=ptp2
.Ss Help
The C utility shows a selection of useful commands when invoked with no parameters.
.Sh BUGS
Please raise bug reports at:
.Aq https://github.com/Xilinx-CNS/sfptpd/issues
.Pp
Support for users of AMD Solarflare NICs is available from:
.Aq support-nic@amd.com
.Pp
It is advisabled to include an archive of
.Pa /var/lib/sfptpd
and the current daemon configuration in any support request.
.Sh AUTHORS
Xilinx, Inc.
.Sh SEE ALSO
.Xr sfptpd 8 .
.Pp
Full documentation is available at:
.Aq https://docs.xilinx.com/r/en-US/ug1602-ptp-user
.Sh COPYRIGHT
Copyright (c) 2022 Xilinx, Inc.


CHANGELOG
Changelog
---------

### v3.7.1.1000 Release Candidate

Features:
 ~ Minor command line enhancements to support container usage (SWPTP-1401)
     - Allow config to be read from stdin with '-f -'
     - Add '--console' option to redirect logs to console

Issue SWPTP-1395
 ~ Fix missing sfptpmon script in .tar.gz release package
 ~ Omit unnecessary new reference to system clock frequency correction state
   file in .rpm packages to avoid confusing 'alien' repackager tool
Issue SWPTP-1403
 ~ Ignore irrelevant PTP packets in sfptpmon event monitoring script instead of
   exiting
Issue SWPTP-1405
 ~ Avoid theoretical risk of double clock correction on startup
 ~ Extend period NIC considered to have unset time to 5 years after epoch
Issue SWPTP-1408
 ~ Fix possible risk of stack corruption in teaming netlink handler


### v3.7.0.1006 Feature Release

Features:
 ~ Allow running as non-root user with minimum required capabilities (SWPTP-442)
     - enable with 'user' configuration or --user command line options
     - root user required for chronyd clock control script to restart chronyd
     - example udev rules support hotplug with non-root user
 ~ Support multiple PPS sources (SWPTP-776)
 ~ Support bridges including over bonds in similar way to LACP bonds (SWPTP-826)
 ~ Support MACVLANs when timestamping enabled in host netns (SWPTP-992)
 ~ Allow a bond, bridge or vlan to be specified for freerun clock (SWPTP-1169)
 ~ Replace bond, team and interface probing with netlink scanning (SWPTP-1269)
 ~ Mix 'efx' clock diff method with list of preferred phc methods (SWPTP-1285)
 ~ Add sync-instances file showing factors influencing selection (SWPTP-1286)
 ~ Add ntp and servo logging modules for targetted diagnostics (SWPTP-1291)
 ~ Reflect external constraints (ntpd, chronyd) as selection rule (SWPTP-1295)
 ~ Better quality logging and trace on hotplug events (SWPTP-1300)
 ~ IP address of parent clock added to state and nodes files (SWPTP-1312)
 ~ Better init system integration with new command line options (SWPTP-1321)
     --daemon       supersedes now-deprecated 'daemon' config file option
     --no-daemon    overrides now-deprecated 'daemon' config file option
     --version      outputs version number
     --test-config  tests configuration file
     --user USER[:GROUP]
     and synchronises with systemd on startup as 'notify'-type service
 ~ Support up to 6 destinations for PTP event monitoring messages (SWPTP-1340)
 ~ Add 'sfptpmon' script to receive PTP event monitoring messages (SWPTP-1342)
 ~ Accept any port for unicast PTP event monitoring destinations (SWPTP-1346)
 ~ The 'ntpsec' implementation of ntpd available in the EPEL repositories for
     RHEL 9 is confirmed to be fully interoperable with sfptpd, allowing
     system clock disciplining to be controlled at runtime.

Removed support:
 ~ Special provision for SFN5000 and SFN6000 series adapters removed (SWPTP-1306)
 ~ Provision for hardware timestamping without Linux PHC removed (SWPTP-1306)
 ~ Support, testing and packaging for RHEL6 is dropped (SWPTP-1371)

Issue SWPTP-1278
 ~ Fix retrieval of transmit timestamps to recover from nic reset quicker
Issue SWPTP-1296
 ~ Improve logging of timer fd activity to identify system configuration issues
Issue SWPTP-1314
 ~ Fix issues using PPS via /dev/ptp external timestamping pin
Issue SWPTP-1326
 ~ Alter buffering for RT JSON stats output to avoid excessive writes per sec
Issue SWPTP-1347
 ~ Remove bogus reserved field from rx timing event monitoring TLV
   ** NOTE ** Faulty senders and receivers need updating together.
              This only applies to users of the mon_rx_sync_timing_data option.
Issue SWPTP-1358
 ~ Fix chronyd client state handling avoiding crashes after system startup
Issue SWPTP-1366
 ~ Fix ntpd client handling to avoid quitting when peer stats not available


### v3.6.0.1015 Feature Release

Features:
 ~ Support Xilinx Alveo X3522 Ethernet adapters (SWPTP-1162)

 ~ Support for operating alongside the chronyd service is now provided in a
    separate 'crny' sync module from the 'ntp' module used for ntpd, e.g.
     [general]
     sync_module crny crny1

    To aid the transition from existing behaviour under the 'ntp' sync module
    a suitable instance is created implicitly. To disable this for simplified
    logging and diagnostics, specify a sync module with no instances:
     [general]
     sync_module crny

 ~ Allow user-defined scripts to reconfigure and restart a chronyd service
    as a workaround to lack of a 'clock control' facility in chronyd. This
    brings parity in the supportable use cases of ntpd and chronyd. (SWPTP-1216)
     [crny]
     control_script /usr/share/doc/sfptpd/examples/chrony_clockcontrol.py

   ** N.B. restarting chronyd is destabilising for NTP synchronisation so this
    feature should be avoided when chrony's time sync performance influences
    sync instance selection. In the absence of native runtime control for
    chronyd it is best used in an always-on or always-off mode with sfptpd. If
    chronyd control is required for fallback it is advisable to give it a
    lower (numerically higher) 'priority' setting so it is only selected on
    total failure of PTP and to increase 'selection_holdoff_period' to 60s.
    Similarly, it would not be recommended to use this clock control method
    while chrony was operating as a clustering determinant (i.e. discriminator).

 ~ Allow operation but block system clock updates when an uncontrollable
    chronyd service is running, except when this condition holds at startup.
    This is improved behaviour when the conflict is transient. It is, however,
    recommended in such situations to configure sfptpd explicitly to treat
    the system clock as read only. To disable this check at startup,
    configure as follows. (SWPTP-1266)
     [general]
     ignore_critical clock-control-conflict

 ~ Allow operation in the absence of a kernel PTP subsystem or PTP and PPS

src/sfptpdctl/sfptpdctl.c
/* SPDX-License-Identifier: BSD-3-Clause */
/* (c) Copyright 2016-2024 Xilinx, Inc. */

/**
 * @file   sfptpdctl.c
 * @brief  Client to control an sfptpd daemon.
 */


#include <stdio.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <stdlib.h>


/****************************************************************************
 * Constant
 ****************************************************************************/

/** sfptpd control socket path */
#define SFPTPD_CONTROL_SOCKET_PATH  "/run/sfptpd/control-v1.sock"

static const char *opts_short = "hs:";
static const struct option opts_long[] = {
	{ "help", 0, NULL, (int) 'h' },
	{ "socket", 1, NULL, (int) 's' },
	{ NULL, 0, NULL, 0 }
};


/****************************************************************************
 * Local Data
 ****************************************************************************/


/****************************************************************************
 * Local functions
 ****************************************************************************/

static void usage(FILE *stream)
{
	fprintf(stream,
		"syntax: %s [OPTIONS] COMMAND-STRING*\n"
		"\n"
		"  COMMAND-STRING\n"
		"    exit                 cause sfptpd to exit\n"
		"    logrotate            cause the log files to be closed and reopened\n"
		"    stepclocks           cause the clocks to be stepped\n"
		"    testmode=MODE[,ARG]* select test mode (see sfptpd source)\n"
		"    selectinstance=NAME  select specific sync instance\n"
		"    dumptables           dump some internal state to message log\n"
		"    pid_adjust=[KP[,[KI][,[KD][,local|ptp|pps|reset]*]]]\n"
		"                         set PID coefficients with optional reset per servo type, or all by default\n"
		"\n"
		"  OPTIONS\n"
		"    -h, --help           Show usage\n"
		"    -s, --socket         Set control socket (default: %s)\n",
		program_invocation_short_name, SFPTPD_CONTROL_SOCKET_PATH);
}


/****************************************************************************
 * Global functions
 ****************************************************************************/

int main(int argc, char *argv[])
{
	const char *control_addr = SFPTPD_CONTROL_SOCKET_PATH;
	struct sockaddr_un addr;
	int control_fd;
	int index;
	int opt;
	int rc;
	int i;

	/* Handle command line arguments */
	while ((opt = getopt_long(argc, argv, opts_short, opts_long, &index)) != -1) {
		switch (opt) {
		case 'h':
			usage(stdout);
			return EXIT_SUCCESS;
		case 's':
			control_addr = optarg;
			break;
		default:
			fprintf(stderr, "unexpected option: %s\n", argv[optind]);
			usage(stderr);
			return EXIT_FAILURE;
		}
	}

	if (optind >= argc) {
		usage(stderr);
		return EXIT_FAILURE;
	}

	/* Write the destination address */
	addr.sun_family = AF_UNIX;
	if (strlen(control_addr) >= sizeof addr.sun_path) {
		fprintf(stderr, "address too long: %s\n", control_addr);
		return EXIT_FAILURE;
	}
	strncpy(addr.sun_path, control_addr, sizeof addr.sun_path);

	/* Create a Unix domain socket for sending control packets */
	control_fd = socket(AF_UNIX, SOCK_DGRAM, 0);
	if (control_fd == -1) {
		perror("socket");
		return EXIT_FAILURE;
	}

	/* Connect to the path in the filesystem. */
	rc = connect(control_fd, (const struct sockaddr *) &addr, sizeof addr);
	if (rc == -1) {
		perror("connect");
		return EXIT_FAILURE;
	}

	/* Send each positional argument as a separate command */
	for (i = optind; i < argc; i++) {
		rc = write(control_fd, argv[i], strlen(argv[i]));
		if (rc != strlen(argv[i])) {
			perror("write");
			return EXIT_FAILURE;
		}
	}

	rc = close(control_fd);
	if (rc == -1) {
		perror("close");
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}
